storage/s3/options.go

```diff
@@ -44,26 +46,158 @@ var (
-	FilenameSuffix *string
 	// ExternalURI is the URI used for signing requests.
 	ExternalURI *string
 	// URI is the URI for the s3 API.
 	URI *string
+	// ProxyURI is used for rewriting presigned requests, pointing the
+	// requests to the proxy URL instead of the direct URL to s3.
+	ProxyURI *url.URL
 
 	// ForcePathStyle encodes bucket in the API path.
 	ForcePathStyle bool
 	// UseAccelerate enables s3 Accelerate
 	UseAccelerate bool
+}
 
+func newFromParent(defaults *storageSettings, parent *model.StorageSettings) *storageSettings {
+	ret := new(storageSettings)
+	ret.patch(defaults)
+	if parent.Bucket != "" {
+		ret.BucketName = &parent.Bucket
+	}
+	if parent.Region != "" {
+		ret.Region = &parent.Region
+	}
+	if parent.Key != "" && parent.Secret != "" {
+		ret.StaticCredentials = &StaticCredentials{
+			Key:    parent.Key,
+			Secret: parent.Secret,
+			Token:  parent.Token,
+		}
+	}
+	if parent.ExternalUri != "" {
+		ret.ExternalURI = &parent.ExternalUri
+	}
+	if parent.Uri != "" {
+		ret.URI = &parent.Uri
+	}
+	if parent.ForcePathStyle != ret.ForcePathStyle {
+		ret.ForcePathStyle = parent.ForcePathStyle
+	}
+	if parent.UseAccelerate != ret.UseAccelerate {
+		ret.UseAccelerate = parent.UseAccelerate
+	}
+	return ret
+}
+
+func (s storageSettings) Validate() error {
+	return validation.ValidateStruct(&s,
+		validation.Field(&s.StaticCredentials),
+	)
+}
+
+func (s storageSettings) options(opts *s3.Options) {
+	if s.StaticCredentials != nil {
+		opts.Credentials = *s.StaticCredentials
+	}
+	if s.Region != nil {
+		opts.Region = *s.Region
+	}
+	if s.URI != nil {
+		endpointURI := *s.URI
+		opts.EndpointResolver = s3.EndpointResolverFromURL(endpointURI,
+			func(ep *aws.Endpoint) {
+				ep.HostnameImmutable = s.ForcePathStyle
+			},
+		)
+	}
+	opts.UsePathStyle = s.ForcePathStyle
+	opts.UseAccelerate = s.UseAccelerate
+}
+
+func (s storageSettings) presignOptions(opts *s3.PresignOptions) {
+	if s.ExternalURI != nil {
+		presignURL := *s.ExternalURI
+		resolver := s3.EndpointResolverFromURL(presignURL,
+			func(ep *aws.Endpoint) {
+				ep.HostnameImmutable = s.ForcePathStyle
+			},
+		)
+		s3.WithPresignClientFromClientOptions(
+			s.options,
+			s3.WithEndpointResolver(resolver),
+		)(opts)
+	}
+}
+
+func (s *storageSettings) patch(setting *storageSettings) *storageSettings {
+	if setting == nil {
+		return s
+	}
+	if setting.BucketName != nil {
+		s.BucketName = setting.BucketName
+	}
+	if setting.StaticCredentials != nil {
+		s.StaticCredentials = setting.StaticCredentials
+	}
+	if setting.Region != nil {
+		s.Region = setting.Region
+	}
+	if setting.ExternalURI != nil {
+		s.ExternalURI = setting.ExternalURI
+	}
+	if setting.URI != nil {
+		s.URI = setting.URI
+	}
+	if setting.ProxyURI != nil {
+		s.ProxyURI = setting.ProxyURI
+	}
+	if setting.ForcePathStyle != s.ForcePathStyle {
+		s.ForcePathStyle = setting.ForcePathStyle
+	}
+	if setting.UseAccelerate != s.UseAccelerate {
+		s.UseAccelerate = setting.UseAccelerate
+	}
```

Comment:
nit/question out of curiosity: why are we comparing here and for the rest we only check !=nil?

Reply:
It's a boolean